src/lib/tls/tls_signature_scheme.cpp

```diff
@@ -0,0 +1,336 @@
+         return { "ECDSA", Botan::EC_Group("secp256r1").DER_encode(Botan::EC_Group_Encoding::NamedCurve) };
+      case ECDSA_SHA384:
+         return { "ECDSA", Botan::EC_Group("secp384r1").DER_encode(Botan::EC_Group_Encoding::NamedCurve) };
+      case ECDSA_SHA512:
+         return { "ECDSA", Botan::EC_Group("secp521r1").DER_encode(Botan::EC_Group_Encoding::NamedCurve) };
+
+      case EDDSA_25519:
+         return { "Ed25519", AlgorithmIdentifier::USE_EMPTY_PARAM };
+
+      case RSA_PKCS1_SHA1:
+      case RSA_PKCS1_SHA256:
+      case RSA_PKCS1_SHA384:
+      case RSA_PKCS1_SHA512:
+      case RSA_PSS_SHA256:
+      case RSA_PSS_SHA384:
+      case RSA_PSS_SHA512:
+         return { "RSA", AlgorithmIdentifier::USE_NULL_PARAM };
+
+      default:
+         return AlgorithmIdentifier();
+      }
+   }
+
+std::optional<Signature_Format> Signature_Scheme::format() const noexcept
+   {
+   switch(m_code)
+      {
+      case RSA_PKCS1_SHA1:
+      case RSA_PKCS1_SHA256:
+      case RSA_PKCS1_SHA384:
+      case RSA_PKCS1_SHA512:
+      case RSA_PSS_SHA256:
+      case RSA_PSS_SHA384:
+      case RSA_PSS_SHA512:
+         return IEEE_1363;
+
+      case ECDSA_SHA1:
+      case ECDSA_SHA256:
+      case ECDSA_SHA384:
+      case ECDSA_SHA512:
+      case EDDSA_25519:
+      case EDDSA_448:
+      case DSA_SHA1:
+      case DSA_SHA256:
+      case DSA_SHA384:
+      case DSA_SHA512:
+         return DER_SEQUENCE;
+
+      default:
+         return std::nullopt;
+      }
+   }
+
+bool Signature_Scheme::is_compatible_with(const Protocol_Version& protocol_version) const noexcept
+   {
+   // RFC 8446 4.4.3:
+   //   The SHA-1 algorithm MUST NOT be used in any signatures of
+   //   CertificateVerify messages.
+   //
+   // Note that Botan enforces that for TLS 1.2 as well.
+   if(hash_function_name() == "SHA-1")
+      return false;
+
+   // RFC 8446 4.4.3:
+   //   RSA signatures MUST use an RSASSA-PSS algorithm, regardless of whether
+   //   RSASSA-PKCS1-v1_5 algorithms appear in "signature_algorithms".
+   //
+   // Note that this is enforced for TLS 1.3 and above only.
+   if(!protocol_version.is_pre_tls_13() &&
+       (m_code == RSA_PKCS1_SHA1   ||
+        m_code == RSA_PKCS1_SHA256 ||
+        m_code == RSA_PKCS1_SHA384 ||
+        m_code == RSA_PKCS1_SHA512))
+      return false;
+
+   return true;
+   }
+
+bool Signature_Scheme::is_suitable_for(const Private_Key &private_key) const noexcept
+   {
+   if(algorithm_name() != private_key.algo_name())
+      return false;
+
+   // The ECDSA private key length must match the utilized hash output length.
+   const auto keylen = private_key.key_length();
+   if(keylen <= 250)
+      return false;
+
+   if(m_code == ECDSA_SHA256 && !(keylen >= 250 && keylen <= 350))
+      return false;
+
+   if(m_code == ECDSA_SHA384 && !(keylen >= 350 && keylen <= 450))
+      return false;
+
+   if(m_code == ECDSA_SHA512 && !(keylen >= 450 && keylen <= 550))
+      return false;
+
+   return true;
+   }
```

Comment:
BoGo tests whether the curve's bit length matches the associated hash function output length. We we weren't sure how to test for this and went with a "fuzzy" range for the ECC key length. The idea was to incorporate custom curves whose bit key length don't fall exactly on predefined thesholds. Does it make sense to implement it like that? @randombit